aya/src/programs/probe.rs

```diff
@@ -48,14 +120,139 @@ pub(crate) fn attach(
         |(_code, io_error)| ProgramError::SyscallError {
             call: "perf_event_open".to_owned(),
             io_error,
         },
     )? as i32;
 
-    perf_attach(program_data, fd)
+    Ok(fd)
+}
+
+fn create_as_trace_point(
+    kind: ProbeKind,
+    name: &str,
+    offset: u64,
+    pid: Option<pid_t>,
+) -> Result<(i32, String), ProgramError> {
+    use ProbeKind::*;
+
+    let (event_type, event_alias) = match kind {
+        KProbe | KRetProbe => (
+            "kprobes",
+            create_probe_event(kind, "kprobe", name, offset)
+                .map_err(|(filename, io_error)| KProbeError::FileError { filename, io_error })?,
+        ),
+        UProbe | URetProbe => (
+            "uprobes",
+            create_probe_event(kind, "uprobe", name, offset)
+                .map_err(|(filename, io_error)| UProbeError::FileError { filename, io_error })?,
+        ),
+    };
+
+    let tpid = read_sys_fs_trace_point_id(event_type, &event_alias)?;
+    let fd = perf_event_open_trace_point(tpid, pid).map_err(|(_code, io_error)| {
+        ProgramError::SyscallError {
+            call: "perf_event_open".to_owned(),
+            io_error,
+        }
+    })? as i32;
+
+    Ok((fd, event_alias))
+}
+
+fn create_probe_event(
+    kind: ProbeKind,
+    event_type: &str,
+    fn_name: &str,
+    offset: u64,
+) -> Result<String, (String, io::Error)> {
+    use ProbeKind::*;
+
+    let events_file_name = format!("/sys/kernel/debug/tracing/{}_events", event_type);
+    let probe_type_prefix = match kind {
+        KProbe | UProbe => 'p',
+        KRetProbe | URetProbe => 'r',
+    };
+    let event_alias = format!("{}_{}_aya_{}", probe_type_prefix, fn_name, process::id());
+
+    let mut events_file = fs::OpenOptions::new()
+        .append(true)
+        .open(&events_file_name)
+        .map_err(|e| (events_file_name.clone(), e))?;
+
+    let p = match kind {
+        KProbe => {
+            if offset > 0 {
+                format!(
+                    "{}:{}s/{} {}+{}",
+                    probe_type_prefix, event_type, event_alias, fn_name, offset
+                )
+            } else {
+                format!(
+                    "{}:{}s/{} {}",
+                    probe_type_prefix, event_type, event_alias, fn_name
+                )
+            }
+        }
+        KRetProbe => format!(
+            "{}:{}s/{} {}",
+            probe_type_prefix, event_type, event_alias, fn_name
+        ),
+        UProbe | URetProbe => {
+            if offset > 0 {
+                format!(
+                    "{}:{}s/{} {}:{:#x}",
+                    probe_type_prefix, event_type, event_alias, fn_name, offset
+                )
+            } else {
+                format!(
+                    "{}:{}s/{} {}",
+                    probe_type_prefix, event_type, event_alias, fn_name
+                )
+            }
+        }
+    };
+
+    events_file
+        .write_all(p.as_bytes())
```

Comment:
do we need to write a line here? because we aren't but then in `find_in_sys_kernel_debug_tracing_events` we iterate lines

Reply:
Writing to the file is how the trace point is deleted. It seems to work (verified with strace) so I'm not sure what you mean by it not writing a line?

Reply:
Sorry I meant we're not adding a final "\n" when we write

Reply:
Oh right, newline vs line :) So the docs here https://www.kernel.org/doc/Documentation/trace/kprobetrace.txt don't seem to indicate either way, but use `echo` in  the examples which would add a new line. So I'm guessing the kernel doesn't rely on it, but it would be more consistent to add a `\n` on write since it would match the expectations from reading line-by-line.